### PR TITLE
ARROW-5848: [C++] SO versioning schema after release 1.0.0

### PR DESCRIPTION
**Pull request by kszucs**
Described by @kou on the mailing list:

> If we may break ABI compatibility each minor version up
> release ("Y" is increased in "X.Y.Z"), we should include
> minor version into SO major version (100, 101 and 102 in the
> following examples):
> 
>   * 1.0.0 -> libarrow.100.0.0
>   * 1.1.0 -> libarrow.101.0.0
>   * 1.2.0 -> libarrow.102.0.0
> 
> If we don't break ABI compatibility each minor version up
> release, we just use the same SO major version (100 in the
> following examples) in 1.0.0:
> 
>   * 1.0.0 -> libarrow.100.0.0
>   * 1.1.0 -> libarrow.100.1.0
>   * 1.2.0 -> libarrow.100.2.0
> 
> I choose 1XX as SO major version because we already use
> 10-14 for SO major version. We should not use them in the
> future to avoid confusion. So I choose 1XX in the above
> examples.

We can change this schema later, but to resolve the CI failures we need a solution for now.

### CODE REVIEW

**Review by xhochy — APPROVED** [review]
+1, quite happy with this. I don't expect us to have a longterm stable ABI yet. Once we have reached this, we should slow down on SO version bumps.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -39,12 +39,14 @@ if(ARROW_VERSION_MAJOR STREQUAL ""
...
 
 # The SO version is also the ABI version
-if(arrow_VERSION_MAJOR STREQUAL "0")
+if(ARROW_VERSION_MAJOR STREQUAL "0")
   # Arrow 0.x.y => SO version is "x", full SO version is "x.y.0"
   set(ARROW_SO_VERSION "${ARROW_VERSION_MINOR}")
   set(ARROW_FULL_SO_VERSION "${ARROW_SO_VERSION}.${ARROW_VERSION_PATCH}.0")
 else()
-  message(FATAL_ERROR "Need to implement SO version generation for Arrow 1.0+")
+  # Arrow 1.x.y => SO version is "10x", full SO version is "x.y.0"
```
Comment (pitrou) [inline cpp/CMakeLists.txt]: Isn't that `full SO version is "10x.y.0"`?

Comment (kszucs) [reply cpp/CMakeLists.txt]: It is, fixing.

**Review by pitrou — APPROVED** [review]
+1, please merge quickly :-)

**Comment by pitrou** [comment]
@kszucs Need to re-run cmake-format, see https://travis-ci.org/kszucs/arrow/jobs/554237034

**Comment by kszucs** [comment]
Ahh, thanks for notifying me.

**Comment by pitrou** [comment]
Can you cancel the old builds on your Travis and AppVeyor account?

**Comment by kszucs** [comment]
Done.

**Comment by pitrou** [comment]
I merged to unstuck other PRs. Thank you!

**Pull request closed without merging**